contracts/MarketQueues.sol

```diff
@@ -0,0 +1,279 @@
+    event EntryRemoved(uint indexed marketId, address indexed channel, bytes32 indexed entryId);
+    event ContractCreated(uint indexed marketId, address newContract, uint notionalValue);
+
+
+    /*
+     *  Data
+     */
+    struct Market {
+        uint id;
+        string name;
+        bool open;
+        address etherChannel;           // ether guy enters queue here
+        address dollarChannel;          // dollar guy enters queue here
+    }
+
+    uint lastMarket = 0;
+
+    // market id to market
+    mapping (uint => Market) markets;
+
+    // map of channels back to market ids
+    mapping (address => uint) channelToMarket;
+
+    // channel address -> Queue of Entries
+    using EntryQueue for EntryQueue.Queue;
+    mapping (address => EntryQueue.Queue) queues;
+
+    address public matcher;             // authorized match maker
+    address public contractStore;
+    address public factory;
+    address public exRate;
+
+
+    /*
+     *  Modifiers
+     */
+    modifier isMatcher() {
+         if (msg.sender != matcher) { throw; }
+         _;
+     }
+
+    modifier registeredChannelAndMarketOpen() {
+        if (markets[channelToMarket[msg.sender]].open != true) { throw; }
+        _;
+    }
+
+
+    /*
+     *  Functions
+     */
+
+    function MarketQueues(address _contractStore, address _factory, address _exRate) {
+        contractStore = _contractStore;
+        factory = _factory;
+        exRate = _exRate;
+        setMatcher(msg.sender);  // for now just make the matcher the creater
+    }
+
+    function setMatcher(address _newMatcher)
+        onlyOwner
+    {
+        matcher = _newMatcher;
+    }
+
+    function createEntry(address _account)
+        external
+        payable
+        registeredChannelAndMarketOpen
+        returns (bytes32 entryId)
+    {
+        address channel = msg.sender;
+        uint marketId = channelToMarket[channel];
+        entryId = sha3(marketId, channel, _account, msg.value, block.number);
+
+        EntryQueue.Queue queue = queues[channel];
+		queue.pushTail(entryId, _account, msg.value);
+
+        EntryAdded(marketId, channel, _account, entryId, msg.value);
+    }
+
+    function getEntryEtherGuy(uint _marketId, bytes32 _entryId)
+        constant
+        returns (
+            address account,
+            uint value,
+            bool filled
+        )
+    {
+        Market market = markets[_marketId];
+        return queues[market.etherChannel].get(_entryId);
+    }
+
+    function getEntryDollar(uint _marketId, bytes32 _entryId)
+        constant
+        returns (
+            address account,
+            uint value,
+            bool filled
+        )
+    {
+        Market market = markets[_marketId];
+        return queues[market.dollarChannel].get(_entryId);
+    }
+
+    function getOpenEtherGuy(uint _marketId)
+        constant
+        returns (bytes32[] ids)
```

Comment:
Can we standardize on "emitter"/"beneficiary" or is that not correct terminology?